Eerste implementatie ADAGUC-checker functionaliteit

saskiawagenaar opened the pull request.

saskiawagenaar committed:
Created first impelementation of the CReporter, which will keep track of all errors, warnings and infomessages.

saskiawagenaar committed:
First usage of the CReporter, currently printing output in de log file.

Commit:
Remove trailing whitespace

Commit:
Remove more trailing whitespace

Commit:
Refactored checkCascadedDimensions

saskiawagenaar committed:
Refactored method for obtaining CRS info from config file + added reporting lines.

saskiawagenaar committed:
Merge branch 'checker' of github.com:saskiawagenaar/adaguc-server into checker

saskiawagenaar committed:
Removed unecessary report statement.

saskiawagenaar committed:
Refactored methods for determining the CRS from the projection variable according to the ADAGUC and the CF-conventions method. Also added all necessary reporting messages.

saskiawagenaar committed:
Refactored method to determine the indices of X and Y dimensions and added the necessary reporting statements.

saskiawagenaar committed:
Added error reporting statement when X and Y vars are not defined for their corresponding dimensions.

saskiawagenaar committed:
Removed unused code.

saskiawagenaar committed:
Refactored method to determine the stride value. Also added reporting statements for the checker.

neutvd committed:
Added reports in CAutoResource.

Changed some methods to be more const correct.
Enabled debug flag in hclasses Makefile.

neutvd committed:
Merge branch 'checker' of https://github.com/saskiawagenaar/adaguc-server into checker

neutvd committed:
Git ignore tags and gdb history.

saskiawagenaar committed:
Refactored methods to determine the X and Y variables and the dWidth and dHeight and added corresponding reporting messages.

saskiawagenaar committed:
Merge branch 'checker' of github.com:saskiawagenaar/adaguc-server into checker

neutvd committed:
Fixed regression in CServerParams.

Introduced in f1af8fd08b4b1239109783ff1c125e6a92089944.
Caused by careless copy/paste.

neutvd committed:
Merge branch 'checker' of github.com:saskiawagenaar/adaguc-server into checker

saskiawagenaar committed:
Refactored method to determine cell size and BBOX and added necessary reporting statements.

neutvd committed:
Write checker reports to separate file

neutvd committed:
Merge branch 'checker' of github.com:saskiawagenaar/adaguc-server into checker

saskiawagenaar committed:
Renamed level2compatmode to formatConverterActive for clarity reasons.

saskiawagenaar committed:
Merge branch 'checker' of github.com:saskiawagenaar/adaguc-server into checker

neutvd committed:
Modified and added some more auto resource report messages

neutvd committed:
Merge branch 'checker' of github.com:saskiawagenaar/adaguc-server into checker

saskiawagenaar committed:
Added reporting statements when reading X and Y data and added some comments.

saskiawagenaar committed:
Merge branch 'checker' of github.com:saskiawagenaar/adaguc-server into checker

neutvd committed:
Set path for report to docker output directory.

neutvd committed:
Merge upstream changes into gitbook updates.

neutvd committed:
Merge branch 'master' into checker

saskiawagenaar committed:
Merge branch 'master' into checker

neutvd committed:
Reporter messages optionally routed to logfile.

- Wrapped CReporter::add*() methods in macros to add filenamme, line
number and class name automatically.
- Replaced the CReporter::add*() calls with these macros.
- Expanced CReporter::add*() calls to optionally take filename, line
number and classname as argument. Set default values.
- Add a flag to turn routing of report messages to logfile. This should
become a command line option at some point.

Users of the CReporter class now need to use the CREPORT_WARN,
CREPORT_ERROR and CREPORT_INFO macros, although using the
CReporter::add*() methods directly shouldn't break anything.

neutvd committed:
Fixed message indicators from CReporter.

For debugging purposes it was nice to be able to grep on what messages
in the log file came from the CReporter class. So they were different.
This commit makes them te same as the indicators from the logging
functions.

neutvd committed:
Revised code comment and fixed indentation.

saskiawagenaar committed:
The checker report is now written in a basic JSON format.

A new header only json parser and serializer has been introduced to facilitate writing and parsing JSON:
https://github.com/nlohmann/json
In order to serialize ADAGUC specific data types, the json_adaguc header and implementation have been added, currently including an implementation to serialize CT::string.
A separate CReportWriter function has been introduced to ensure that only this writer has a dependency to the new JSON library,
not the CReporter class itself, preventing long compilation times since CReporter is used in many different ADAGUC files.

saskiawagenaar committed:
Added a const char* conversion operator to the CT::string class.
Now it is not necessary to call c_str() when an implicit conversion to const char* is possible.

saskiawagenaar committed:
Changed the checker report path back to the original value instead of a test value.

Commit:
Update Dockerfile and compile.sh to enable building nlohmann/json

Problem: nlohmann/json requires c++11 which isn't supported by default gcc on centos:7
Solution: Use centos/devtoolset-7-toolchain-centos7:7 instead, which includes gcc-7.

Also run as USER root because centos/devtoolset-7-coolchain-centos7 runs as USER default
and install openssl which is missing.

saskiawagenaar committed:
Added categories and documentation links to the report messages.

Introduced a new class CReportMessage to hold all information of a report message,
including the category and the documentation link. Updated the macros for logging
such that it is possible to pass a category and a documentation link. Also updated
the formatting of the Json output, such that all information concerning a report
message is logged. The order of the messages is now also preserved.

saskiawagenaar committed:
Removed the 'FATAL' severity option for reporting messages.

ADAGUC keeps track of the existences of logged error messages and considers it fatal sometimes.
Also refactored all occurences of the 'ERROR' severity which were in fact a 'WARNING'.
Removed all duplicated CDBERROR calls.

saskiawagenaar committed:
Deduplicated debug messages and re-added debug logging for info report messages.

saskiawagenaar committed:
Removed -g option in default compiler settings and added std=c++11 in all other cases.

saskiawagenaar committed:
Made method CReporter.writeMessageToLog const correct.

saskiawagenaar committed:
Merge pull request #4 from saskiawagenaar/EUT-144-crash

Added categories and documentation links and formatted json output checker report.

saskiawagenaar committed:
Merge pull request #1 from saskiawagenaar/EUT-144

Checker report written as JSON

neutvd committed:
Add report messages to CProj4ToCF class.

- Added report messages to give hints on chosen projection.
- Removed large white space blocks and commented out code.

neutvd committed:
Added report messages to CProj4ToCF.

- Refactored the units setting in CProj4ToCF::convertCFToProj into a
  member method.
- Added const correctness to CCDFVariable::getParentCDFObject so we can
  make CProj4ToCF::getProjectionUnits and CProj4ToCF::getProjectionUnits
  const correct.
- Fixed linker error in Makefile resulting from having to include
  CReporter.h from CCDFDataModel library.

neutvd committed:
Removed "FATAL" sevirity string.

ERRORs are always fatal, so no need to have a separate severity level
for it.

neutvd committed:
Report "no time units", rather than log it.

A file issued by WUR didn't visualize and also didn't generate a
report message indicating the problem.
TODO:
- CDBFileScanner.cpp has several more locations where a report
message could be helpful.
- Write unit test for this file.

neutvd committed:
Added unit test for testing absence of time unit.

- Test if a report is generated.
- Test if the correct error message is reported.
- Test if the report messages have all mandatory keys.
- Test if adaguc exits with exit code 1.

neutvd committed:
Added checkReport() to test for correct report.

- Called from test_WMSGetCapabilities_testdatanc and
test_WMSGetMap_testdatanc only for now.
TODO:
- When checker command line option code merged in, make sure the
report ends up in the test's temporary directory so it is always
removed before performing the test.
- Add more expected report files and call checkReport() from more
tests.

neutvd committed:
Added more report messages to CDBFileScanner.

Don't seem to be touched by adaguc-checker at this point, but
they are good candidates for reporting.

saskiawagenaar committed:
Added an environment variable which allows for specifying the location of the checker report.

neutvd committed:
Merge branch 'master' into checker

Merged the command line option parsing updates on master with the
checker branch (as well as some other bug fixes). Now it should be
possible to add a command line option to set the report filename.

neutvd committed:
EUT-202 - Adds command line option for reports.

- Use --report to write report file in default location.
- Use --report=filename to write report in filename
- When neither is given and ADAGUC_CHECKER_FILE environment variable is
set, write report file to the value of ADAGUC_CHECKER_FILE environment
variable.
- When neither is given and ADAGUC_CHECKER_FILE is not set, no report is
written.

TestWMS.test_WMSGetMap_testdatanc tests the first option.
TestWMS.test_WMSGetMap_Report_nounits tests the second option.
TestWMS.test_WMSGetCapabilities_testdatanc tests the last option.

TODO: write a test for the third (set by environment) option.

jorisknmi committed:
EUT-202 - Command line option for reports; added unittest test_WMSGetMap_Report_env

- Use --report=filename to write report specified with filename
- Use --report          to write report specified in environment variable ADAGUC_CHECKER_FILE
  if ADAGUC_CHECKER_FILE is not set, write file to default location
- If --report is not given, no report will be written

TestWMS.test_WMSGetMap_Report_nounits tests reporting to a parameter specified location
TestWMS.test_WMSGetMap_Report_env tests reporting to a location specified by environment variable ADAGUC_CHECKER_FILE
TestWMS.test_WMSGetMap_testdatanc tests reporting to default location
TestWMS.test_WMSGetCapabilities_testdatanc tests the option of writing no report

neutvd committed:
Merge pull request #10 from saskiawagenaar/EUT-202

EUT-202 - Adds command line option for reports.

maartenplieger commented on CCDFDataModel/CProj4ToCF.cpp:
```diff
@@ -447,16 +460,19 @@ int CProj4ToCF::convertProjToCF( CDF::Variable *projectionVariable, const char *
   try{
     CT::string grid_mapping_name;
     projectionVariable->getAttribute("grid_mapping_name")->getDataAsString(&grid_mapping_name);
+    CREPORT_INFO_NODOC(CT::string("Unsupported projection. Using CF conventions to determine projection. grid_mapping_name variable: ")+grid_mapping_name,
```
Rephrase to something like: "Unable to get projection from proj4 string, trying  to use CF convention to determine projection. grid_mapping variable : " etc.

maartenplieger commented on tests/AdagucTests/TestWMS.py:
```diff
@@ -231,3 +257,28 @@ def test_WMSGetFeatureInfo_timeseries_forecastreferencetime_json(self):
+        filename="test_WMSGetMap_Report_nounits"
+        reportfilename="./nounits_checker_report.txt"
+        status,data,headers = AdagucTestTools().runADAGUCServer(
+            "source=testdata_report_nounits.nc&service=WMS&request=GetMap&version=1.3.0&layers=sow_a1&crs=EPSG%3A4326&bbox=47.80599631376197%2C1.4162628389784275%2C56.548995855839685%2C9.526486675156528&width=863&height=981&format=image%2Fpng&info_format=application%2Fjson&time=1000-01-01T00%3A00%3A00Z%2F3000-01-01T00%3A00%3A00Z&dim_reference_time=2017-12-15T09%3A00%3A00Z",
```
Please move these files inside a folder which explicitly mark them as invalid datasets for testing purposes

maartenplieger commented on Docker/adaguc-services-config.xml:
```diff
@@ -16,6 +16,7 @@
     <export>ADAGUC_LOGFILE=/var/log/adaguc/adaguc-server.log</export>
     <export>ADAGUC_FONT=/adaguc/adaguc-server-master/data/fonts/FreeSans.ttf</export>
     <export>ADAGUC_ONLINERESOURCE={ENV.EXTERNALADDRESS}/adaguc-services/adagucserver?</export>
+    <export>ADAGUC_CHECKER_DIR={ENV.ADAGUC_CHECKER_DIR}</export>
```
Can be removed, not used anywhere.

neutvd replied:
Done. Commit e7df138088b790ecef9defa310741b7615288120

maartenplieger commented on hclasses/Makefile:
```diff
@@ -5,9 +5,9 @@ ifndef BUILDER_ADAGUCCOMPILERSETTINGS
 endif
 
-CCOMPILER=g++ $(BUILDER_ADAGUCCOMPILERSETTINGS) -I $(INCLUDEDIR)
+CCOMPILER=g++ -g $(BUILDER_ADAGUCCOMPILERSETTINGS) -I $(INCLUDEDIR)
```
Please remove -g, already set via compile.sh in dev mode

maartenplieger replied:
Maybe add -g by default to Makefiles in hclasses and CCDFDataModel

maartenplieger commented on adagucserverEC/CServerParams.cpp:
```diff
@@ -284,14 +283,10 @@ int CServerParams::makeLayerGroupName(CT::string *groupName,CServerConfig::XMLE_
+bool CServerParams::isAutoResourceCacheEnabled() const {
+
+  if (cfg->AutoResource.size() > 0)
+    return cfg->AutoResource[0]->attr.enablecache.equals("true");
```
Previously at first a string compare and consecutive calls result only in an int compare. Not it is always a string compare, why?

neutvd replied:
This was done so that this method could be made const. In the previous version private class member autoResourceCacheEnabled can be changed preventing this method from being changed to const.
I'm kind of hard pressed to believe this will have measurable impact, but we can change the line to

```
return cfg->AutoResource[0]->attr.enablecache.equals("true", 4);
```
Which will compare the length of the strings first. Since "false" is 5 bytes long, this will reduce the number of instructions from the original version if the value is usually "false". If the value is usually true, we can reverse the call:
```
return ! cfg->AutoResource[0]->attr.enablecache.equals("false", 5);
```

Personally, I think this kind of optimization is not worth the bother of reducing readability and robustness.

maartenplieger commented on adagucserverEC/CImageDataWriter.cpp:
```diff
@@ -3356,10 +3357,7 @@ int CImageDataWriter::end(){
   if(errorsOccured()){
-      
-    
-    CDBError("Errors occured during image data writing");
```
Are errors printed CREPORT_ERROR_NODOC shown in ADAGUC error reporting if report functionality is not enabled? E.g. is older behavior unchanged?

neutvd replied:
Yes. Wrote a test to verify. Commit a484eb4677fe318b57decea33b506f547f31cebe

maartenplieger commented on hclasses/CReporter.cpp:
```diff
@@ -0,0 +1,73 @@
+                           int line, const char *className) {
+
+  if(this->writelog) {
+    writeMessageToLog(message, severity, file, line, className);
```
Are errors really shown in default ADAGUC logs (without reporting enabled), also see removal of CImageDataWriter.cpp, 3361 (CDBError)

neutvd replied:
Yes. Wrote a test to verify. Commit a484eb4677fe318b57decea33b506f547f31cebe

maartenplieger commented on tests/AdagucTests/TestWMS.py:
```diff
@@ -55,15 +62,34 @@ def test_WMSGetCapabilities_testdatanc(self):
+        self.checkReport(reportFilename="checker_report.txt",
+                         expectedReportFilename="checker_report_WMSGetMap_testdatanc.txt")
+
+    def test_WMSGetMap_Report_env(self):
```
Maybe we can also add a test to check if errors logged with either CDBError or CREPORT_ERROR_NODOC show up in the logs when reporter functionality is disabled.

neutvd replied:
Done. Commit a484eb4677fe318b57decea33b506f547f31cebe

maartenplieger commented on Dockerfile:
```diff
@@ -1,11 +1,15 @@
-FROM centos:7
+FROM centos/devtoolset-7-toolchain-centos7:7
```
Dockerfile should be:

```
FROM centos/devtoolset-7-toolchain-centos7:7
USER root

MAINTAINER Adaguc Team at KNMI <adaguc@knmi.nl>

######### First stage (build) ############

# production packages, same as stage two
RUN yum update -y && yum install -y \
    epel-release deltarpm

RUN yum install -y \
    python-lxml \
    cairo \
    curl \
    gd \
    gdal \
    hdf5 \
    libxml2 \
    logrotate \
    postgresql-server \
    proj \
    tomcat \
    udunits2 \
    openssl \
    netcdf

# building / development packages
RUN yum update -y && yum clean all
RUN yum install -y centos-release-scl && yum install -y devtoolset-7-gcc-c++ && source /opt/rh/devtoolset-7/enable
RUN yum install -y \
    cairo-devel \
    curl-devel \
    gd-devel \
    gdal-devel \
    hdf5-devel \
    libxml2-devel \
    make \
    netcdf-devel \
    openssl \
    postgresql-devel \
    proj-devel \
    sqlite-devel \
    udunits2-devel 

# Install adaguc-server from context
WORKDIR /adaguc
COPY . /adaguc/adaguc-server-master

# Alternatively install adaguc from github
# WORKDIR /adaguc
# ADD https://github.com/KNMI/adaguc-server/archive/master.tar.gz /adaguc/adaguc-server-master.tar.gz
# RUN tar -xzvf adaguc-server-master.tar.gz

WORKDIR /adaguc/adaguc-server-master
RUN bash compile.sh

######### Second stage (production) ############
FROM centos:7

# production packages, same as stage one
RUN yum update -y && yum install -y \
    epel-release deltarpm

RUN yum install -y \
    python-lxml \
    cairo \
    curl \
    gd \
    gdal \
    hdf5 \
    libxml2 \
    logrotate \
    postgresql-server \
    proj \
    tomcat \
    udunits2 \
    openssl \
    netcdf

WORKDIR /adaguc/adaguc-server-master

# Install adaguc-services (spring boot application for running adaguc-server)
RUN curl -L https://jitpack.io/com/github/KNMI/adaguc-services/1.0.4/adaguc-services-1.0.4.war > /usr/share/tomcat/webapps/adaguc-services.war
   
# Install compiled adaguc binaries from stage one    
COPY --from=0 /adaguc/adaguc-server-master/bin /adaguc/adaguc-server-master/bin
COPY --from=0 /adaguc/adaguc-server-master/data /adaguc/adaguc-server-master/data
COPY --from=0 /adaguc/adaguc-server-master/tests /adaguc/adaguc-server-master/tests
COPY --from=0 /adaguc/adaguc-server-master/runtests.sh /adaguc/adaguc-server-master/runtests.sh

# Run adaguc-server functional tests
RUN bash runtests.sh

# Setup directories
RUN mkdir -p /data/adaguc-autowms && \
    mkdir -p /data/adaguc-datasets && \
    mkdir -p /data/adaguc-data && \
    mkdir -p /adaguc/userworkspace && \
    mkdir -p /data/adaguc-services-home && \
    mkdir -p /adaguc/basedir && \
    mkdir -p /var/log/adaguc && \
    mkdir -p /adaguc/adagucdb && \
    mkdir -p /adaguc/security && \
    mkdir -p /data/adaguc-datasets-internal && \
    mkdir -p /servicehealth
    
# Configure
COPY ./Docker/adaguc-server-config.xml /adaguc/adaguc-server-config.xml
COPY ./Docker/adaguc-services-config.xml /adaguc/adaguc-services-config.xml
COPY ./Docker/start.sh /adaguc/
COPY ./Docker/adaguc-server-logrotate /etc/logrotate.d/adaguc
COPY ./Docker/adaguc-server-*.sh /adaguc/
COPY ./Docker/baselayers.xml /data/adaguc-datasets-internal/baselayers.xml
COPY ./Docker/tomcat-server.xml /etc/tomcat/server.xml
RUN  chmod +x /adaguc/adaguc-server-*.sh && chmod +x /adaguc/start.sh

# Set adaguc-services configuration file
ENV ADAGUC_SERVICES_CONFIG=/adaguc/adaguc-services-config.xml 
ENV ADAGUCDB=/adaguc/adagucdb
ENV EXTERNALADDRESS="http://localhost:8080/"

# These volumes are configured in /adaguc/adaguc-server-config.xml
# Place your netcdfs, HDF5 and GeoJSONS here, they will be visualized with the source=<file> KVP via the URI
VOLUME /data/adaguc-autowms   
# Place your dataset XML configuration here, they will be accessible with the dataset=<dataset basename> KVP via the URI
VOLUME /data/adaguc-datasets  
# Place your netcdfs, HDF5 and GeoJSONS here you don't want to have accessible via dataset configurations.
VOLUME /data/adaguc-data      
# Loggings are save here, including logrotate
VOLUME /var/log/adaguc/       
# You can make the postgresql database persistent by externally mounting it. Database will be initialized if directory is empty.
VOLUME /adaguc/adagucdb       
# Settings for HTTPS / SSL can be set via keystore and truststore. Self signed cert will be created if nothing is provided.
VOLUME /adaguc/security

# For HTTP
EXPOSE 8080 
# For HTTPS
EXPOSE 8443 

ENTRYPOINT /adaguc/start.sh
```

neutvd replied:
Committed in e7df138088b790ecef9defa310741b7615288120

neutvd committed:
Rephrases unsupported proj4 string report.

Review comment: https://github.com/KNMI/adaguc-server/pull/66#pullrequestreview-133535342

Make it clear that there's no unsupported projection per se, but that it
couldn't be determined it from the proj4string and it will be attempted
to obtain it from the CF conventions grid_mapping variable.

neutvd committed:
Allow hclasses -g compile flag be set by compile script.

Review comment: https://github.com/KNMI/adaguc-server/pull/66#pullrequestreview-133557672

Set -g flag by default, when make is called from within hclasses
directory, but allow compile.sh master build script to override.

neutvd committed:
Move data for testing bad file behaviour.

testdata_report_nounits.nc is an invalid file missing units, to test
if adaguc generates correct reporting on such a file. Move this file
to a location separate from valid data.
See also https://github.com/KNMI/adaguc-server/pull/66/files/b86267cd0c410e8606382b55efcc7a4f2b4c4756#r199436636

neutvd committed:
Add test checking if errors end up in log file.

There was some doubt expressed in the merge request comments as to
whether messages reported with CREPORT_ERROR_NODOC are also logged to
the normal log file when reporting is turned off.

See https://github.com/KNMI/adaguc-server/pull/66/files#r199443944 and
https://github.com/KNMI/adaguc-server/pull/66/files#r199445156
This test doesn't test that particular behaviour. It tests whether
errors end up in the logfile when reporting is turned on, but this test
can serve as basis for testing the case when reporting is turned off.

neutvd committed:
Adds tests checking if errors in logfile with report off.

See previouse commit 3c474eff07dc21d7b56af222bf353c5af7923e14.
The added test tests the requested behaviour if error messages still end
up in the normal logfile when reporting is turned off.

See pull request comments at
https://github.com/KNMI/adaguc-server/pull/66/files#r199443944 and
https://github.com/KNMI/adaguc-server/pull/66/files#r199445156

neutvd committed:
Solve Dockerfile merge conflict.

Also removes an environment variable that is no longer used.

See pull request: https://github.com/KNMI/adaguc-server/pull/66 comment
https://github.com/KNMI/adaguc-server/pull/66/files#r199460421 and
https://github.com/KNMI/adaguc-server/pull/66/files#r199436962

maartenplieger committed:
Merge branch 'master' into checker
